airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/SecretsRepositoryWriter.java

```diff
@@ -0,0 +1,268 @@
+import io.airbyte.config.DestinationConnection;
+import io.airbyte.config.SourceConnection;
+import io.airbyte.config.StandardDestinationDefinition;
+import io.airbyte.config.StandardSourceDefinition;
+import io.airbyte.config.persistence.split_secrets.SecretPersistence;
+import io.airbyte.config.persistence.split_secrets.SecretsHelpers;
+import io.airbyte.config.persistence.split_secrets.SplitSecretConfig;
+import io.airbyte.protocol.models.ConnectorSpecification;
+import io.airbyte.validation.json.JsonSchemaValidator;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class takes secrets as arguments but never returns a secrets as return values (even the ones
+ * that are passed in as arguments). It is responsible for writing connector secrets to the correct
+ * secrets store and then making sure the remainder of the configuration is written to the Config
+ * Database.
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+public class SecretsRepositoryWriter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SecretsRepositoryWriter.class);
+
+  private static final UUID NO_WORKSPACE = UUID.fromString("00000000-0000-0000-0000-000000000000");
+
+  private final ConfigRepository configRepository;
+  private final JsonSchemaValidator validator;
+  private final Optional<SecretPersistence> longLivedSecretPersistence;
+  private final Optional<SecretPersistence> ephemeralSecretPersistence;
+
+  public SecretsRepositoryWriter(final ConfigRepository configRepository,
+                                 final Optional<SecretPersistence> longLivedSecretPersistence,
+                                 final Optional<SecretPersistence> ephemeralSecretPersistence) {
+    this(configRepository, new JsonSchemaValidator(), longLivedSecretPersistence, ephemeralSecretPersistence);
+  }
+
+  @VisibleForTesting
+  SecretsRepositoryWriter(final ConfigRepository configRepository,
+                          final JsonSchemaValidator validator,
+                          final Optional<SecretPersistence> longLivedSecretPersistence,
+                          final Optional<SecretPersistence> ephemeralSecretPersistence) {
+    this.configRepository = configRepository;
+    this.validator = validator;
+    this.longLivedSecretPersistence = longLivedSecretPersistence;
+    this.ephemeralSecretPersistence = ephemeralSecretPersistence;
+  }
+
+  private Optional<SourceConnection> getOptionalSourceConnection(final UUID sourceId) throws JsonValidationException, IOException {
+    try {
+      return Optional.of(configRepository.getSourceConnection(sourceId));
+    } catch (final ConfigNotFoundException e) {
+      return Optional.empty();
```

Comment:
Why this is not an error? Is there a valid use case where a source don't have a config?

Reply:
it is possible that it hasn't been persisted to the db yet. if it hasn't then this returns empty.